diff --git a/components/spi_flash/spi_flash_os_func_app.c b/components/spi_flash/spi_flash_os_func_app.c
--- a/components/spi_flash/spi_flash_os_func_app.c
+++ b/components/spi_flash/spi_flash_os_func_app.c
@@ -221,4 +221,4 @@ static void* get_buffer_malloc(void* arg, size_t reqest_size, size_t* out_size)
     void* ret = NULL;
-    size_t read_chunk_size = (reqest_size + 3) & ~3;
-
+    size_t read_chunk_size = MAX((reqest_size + 3) & ~3, 64);
+
     while (ret == NULL && read_chunk_size >= 64) {
