diff --git a/subsys/bluetooth/host/Kconfig b/subsys/bluetooth/host/Kconfig
index 4ed05bbda60..d98619f9a95 100644
--- a/subsys/bluetooth/host/Kconfig
+++ b/subsys/bluetooth/host/Kconfig
@@ -15,9 +15,9 @@ config BT_LONG_WQ_STACK_SIZE
 	# Hidden: Long workqueue stack size. Should be derived from system
 	# requirements.
 	int
-	default 1400 if BT_SEND_ECC_EMULATION
-	default 1300 if BT_GATT_CACHING
-	default 1024
+	default 2800 if BT_SEND_ECC_EMULATION
+	default 2600 if BT_GATT_CACHING
+	default 2048

 config BT_LONG_WQ_PRIO
 	int "Long workqueue priority. Should be pre-emptible."

diff --git a/soc/espressif/esp32c3/memory.h b/soc/espressif/esp32c3/memory.h
index 97b2068bb45..784b0a3355e 100644
--- a/soc/espressif/esp32c3/memory.h
+++ b/soc/espressif/esp32c3/memory.h
@@ -46,8 +46,8 @@
 /* For safety margin between bootloader data section and startup stacks */
 #define BOOTLOADER_STACK_OVERHEAD      0x0
 /* These lengths can be adjusted, if necessary: */
-#define BOOTLOADER_DRAM_SEG_LEN        0x9800
-#define BOOTLOADER_IRAM_SEG_LEN        0x9800
+#define BOOTLOADER_DRAM_SEG_LEN        0xb400
+#define BOOTLOADER_IRAM_SEG_LEN        0xb400
 #define BOOTLOADER_IRAM_LOADER_SEG_LEN 0x1400

 /* Start of the lower region is determined by region size and the end of the higher region */
